resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -348,14 +370,49 @@ private[spark] class KubernetesClusterSchedulerBackend(
             .endSpec()
         }
       }.getOrElse(basePodBuilder)
-    val resolvedExecutorPod = executorInitContainerBootstrap.map { bootstrap =>
-      bootstrap.bootstrapInitContainerAndVolumes(
-        "executor",
-        withMaybeShuffleConfigPodBuilder)
-    }.getOrElse(withMaybeShuffleConfigPodBuilder)
+    val executorInitContainerPodBuilder = executorInitContainerBootstrap.map {
+        bootstrap =>
+          bootstrap.bootstrapInitContainerAndVolumes(
+            "executor",
+            withMaybeShuffleConfigPodBuilder)
+      }.getOrElse(withMaybeShuffleConfigPodBuilder)
+
+    val resolvedExecutorPodBuilder = if (nodeToTaskCount.nonEmpty) {
+      val taskTotal = nodeToTaskCount.foldLeft(0)(_ + _._2)
+      // Normalize to node affinity weights in 1 to 100 range.
+      val nodeToWeight = nodeToTaskCount.map{
+        case (node, taskCount) =>
+          (node, math.min(100, math.max(1, math.round(taskCount * 100.0 / taskTotal))))}
+      val weightToNodes = nodeToWeight.values.map(
+        weight => (weight, nodeToWeight.keys.filter(nodeToWeight(_) == weight))).toMap
+      // TODO: Use non-annotation syntax when we switch to K8s version 1.6.
+      val nodeAffinity = s"""{
```

Comment:
Is `NodeAffinity` a construct that isn't supported by the Fabric8 client? It might be better to add an object model there that has a serializer to and from YAML/JSON that we can then add here.

To unblock us for now we can build the POJO class in this project with the appropriate Jackson annotations that allow it to be serialized and deserialized with an `ObjectMapper`.